### ISSUE TITLE
Support self-fetching parameters in flake.nix

### ISSUE THREAD

**Issue by nrdxp** [NixOS/nix#5312]
#5284 happened, so instead I'll open a proper issue to address what I feel would be the "correct" solution since I am incapable of implementing it atm.

I see no particular reason why `self` should be _especially handicapped_ from the other inputs, but should be able to be easily handled as a regular `git` type input, or whichever active VCS Nix can detect.

you wouldn't even have to change the syntax, but simply allow `inputs.self.submodules = true`, for example.

Nix would simply have to know that `inputs.self` is the current `self`.

**Comment by edolstra** [comment]
There is a very good reason why `self` is different: we have to fetch the `self` flake before we can read its `flake.nix`. So any attributes like `submodules` have to be known in advance.

**Comment by nrdxp** [comment]
There is no possible mechanism to read the flake inputs before `self` is fetched? Like maybe just pulling the _lock file_ into the Nix store for reading before the rest of the flake?

**Comment by mstone** [comment]
@nrdxp — I’ve been thinking about this a bit because it’s going to bite everyone who tries to flakeify upstream projects that use submodules, like, in my case, qemu.

So far the best idea I’ve come up with is to write the flake to use an auxiliary input, in addition to `self`, that fetches the same underlying repo just with submodules enabled.

Unfortunately, that still leaves several quality-of-life problems that I wonder if we can fix:

a) that the self input is fetched but is unused (other than for obtaining flake.nix/flake.lock)

b) that the self-with-submodules input is “locked” when we’d rather that it be “derived from” or “tied” to self — e.g. so that we see dirty working copy edits / whatever the flake-ref resolves to when being built, and don’t have to remember to push commits that bump flake.lock just to test changes.

(P.S. — if you’re interested in this approach, I wrote more at https://gist.github.com/mstone/4218e6fa9ef98e5153f543fb8e7b4743 with several variations in this idea, though I’m not really sure that any of them are ready for prime time. :-/)

**Comment by Kha** [comment]
> There is a very good reason why `self` is different: we have to fetch the `self` flake before we can read its `flake.nix`. So any attributes like `submodules` have to be known in advance.

Would this still be the case with #3121? Afaiu, in that case we would evaluate `flake.nix` in the local checkout/git cache and only then copy `self` to the store.

**Comment by edolstra** [comment]
#3121 only addresses the case of a local git tree.

The fundamental issue is that we do a `fetchTree` to get the top-level flake, which requires having all the attributes needed for fetching (like `submodules`). If the top-level `flake.nix` can then change those attributes, we may have to do another `fetchTree`. Also the semantics get pretty ugly, since you could do something like `inputs.self.url = some-arbitrary-url`.

**Comment by kero0** [comment]
Am I missing something? Or why can't the answer just be as simple as manipulate self with command line args? So have a `--fetch-submodules` arg and add more as they are requested.

**Comment by nrdxp** [comment]
Command line args would work in theory, but if I'm conceptualizing it correctly, it would also break purity since the end result would be vastily different depending on the args past.

**Comment by mstone** [comment]
I disagree: — while I think CLI is good (and would likely first pursue implementing it as ?-arguments on `.` for data that conceptually modify the flake rather than the flake-interpreter), I think the point here is that flake-authors need a way to specify all build instructions, including *what code to build*, as part of the flake definition — that way, end-users get something that “just works” without more detailed communication from the flake authors (e.g. of a special URL to build, or extra CLI arguments) whenever possible.

**Comment by nrdxp** [comment]
IIRC, more general flake arguments passed from the cli to the flake builder were considered early on, similar to how one can pass arguments to nix expressions with the old `nix-build`, but were never implemented because of the same purity concerns. So, unless I misunderstood, I dont' see cli args being accepted as a solution.

**Comment by L-as** [comment]
My suggestion for how to fix this:

Flakes could have a `metadata` attribute set in `flake.nix`, as such:
```nix
{
metadata.fetch-git-submodules = true;
}
```

The metadata attributes would also have to be in normal-form, and would not allow any path literals.

This means that Nix can fetch the flake, check the metadata, then check whether to also fetch the submodules *before* evaluating any Nix code.
The metadata could also be in a separate JSON file if that is better.

With this you don't have the double-fetching problem, but it's restricted to Git. This isn't "clean" IMO, but IMHO the problem is Git itself not always fetching submodules. Git should have from the beginning enforced fetching submodules, and we have to work around that now.

The only downside with always fetching submodules in Nix is that it breaks backward compatibility in some subtle ways, and doesn't work well with git-crypt.

**Comment by marksisson** [comment]
Currently, I use `nix build .?submodules=1` to build a flake that uses git submodules.

What if we added a nix configuration option, e.g. `flake-reference-use-submodule`, to set the flake reference submodule attribute to a desired default (if not specified, it could be `false` by default to maintain backwards compatibility).  Then we would not have to pass it on the command line.

Also, since flakes have a `nixConfig` attribute, maybe this configuration option could also be specified in the `flake.nix` to override the option that is specified in `nix.conf`?

**Comment by nrdxp** [comment]
> Also, since flakes have a `nixConfig` attribute, maybe this nix configuration option could also be specified in the `flake.nix` to override the option that is specified in `nix.conf`? or vice-versa?

This sounds like the cleanest solution to me, assuming we can get approval for the new nix.conf setting.

**Comment by roberth** [comment]
`nixConfig` would also affect inputs. We need it to be more fine grained; per-flake. Something like @L-as' suggestion or `fetchConfig` would be more appropriate for this, or `inputs.self` could be special cased, so as to provide a consistent interface.

> With this you don't have the double-fetching problem, but it's restricted to Git.

We have multiple solutions to prevent double fetching.

Caching:
 - For inputs, only double fetch when an input is first added. Save the fetch config to the lock. Reuse this info and double fetch only when the fetch config has changed.
 - For the cli, cache the fetch config after first fetch, just like we do in the lock.

Fetcher dependent solutions:
 - For git, which is most flakes, do a minimal fetch first and expand on that as required by the fetch config.
 - For github, gitlab, etc: use the API to fetch just `flake.nix` first. This is hardly a double fetch, but not as efficient as the previous solutions. Worth keeping in mind though.

The caching solutions introduce significant complexity. This may be worth embracing at some point, but until then, the fetcher dependent solutions don't impose complexity on the rest of the system, making them feasible to implement first, to solve 90% of the problem.

----

I'll rename this issue, as the current problem is only about fetch parameters. The remainder of original problem statement is to treat similar things similarly, and I think we can agree on that. If there's another issue related to `self`, a new issue with a new thread is more productive.

**Comment by roberth** [comment]
Oh well, looks like "Triage" access doesn't include renaming. @nrdxp could you rename this issue to something like "Support self-fetching parameters in flake.nix"?
